app/code/Magento/Catalog/Model/Product/Attribute/Repository.php

```diff
@@ -218,8 +219,8 @@ public function getCustomAttributesMetadata($dataObjectClassName = null)
      */
     protected function generateCode($label)
     {
-        $code = substr(preg_replace('/[^a-z_0-9]/', '_', $this->filterManager->translitUrl($label)), 0, 30);
-        $validatorAttrCode = new \Zend_Validate_Regex(['pattern' => '/^[a-z][a-z_0-9]{0,29}[a-z0-9]$/']);
+        $code = substr(preg_replace('/[^a-zA-Z_0-9]/', '_', $this->filterManager->translitUrl($label)), 0, Attribute::ATTRIBUTE_CODE_MAX_LENGTH);
+        $validatorAttrCode = new \Zend_Validate_Regex(['pattern' => '/^[a-zA-Z][a-zA-Z_0-9]{0,29}[a-zA-Z0-9]$/']);    
```

Comment:
I believe we should accept just lower case characters here, as described on the docs:

https://docs.magento.com/user-guide/stores/attributes-product.html
```
Attribute Code | (Required) A unique identifier for internal use. The Attribute Code must begin with a letter, but can include a combination of lowercase letters (a-z) and numbers (0-9). The code must be less than thirty characters in length and cannot include any special characters or spaces, although an underscore (_) can be used to indicate a space.
```

Reply:
you are right with this lowercase rule cited in the docs, I missed it previously sorry, is there any technical reason to do so btw?  In my particular case the attribute code is created from a management system that stores attribute references as GUID and those have upper case letters and changing the rule here could avoid some headaches.

Reply:
done

Reply:
To be honest, I'm not 100% sure what is the reason for that, but I guess it is to simplify the field itself since it is used as the key for the attribute. Thanks for reverting it.